history-network.md

```diff
@@ -60,10 +60,19 @@ Each of these data items are represented as a key-value pair. Denote the key for
 
 All `content` items are transmitted as RLP-encoded byte arrays.
 
-We derive a `content-id` from the `content-key` as `H(content-key)` where `H` denotes the SHA-256 hash function, which outputs 32-byte values. The `content-id` represents the key in the DHT that we use for `distance` calculations.
-
 All `content-key` values are encoded and decoded according to SSZ sedes.
 
+In addition to the `content-key`, each piece of `content` has a corresponding `content-id`. The `content-id` is the 32 byte value that we use (along with the `node-id`) for `distance` calculations.
+
+We derive a `content-id` from the `content-key` as follows:
+
+  1.) If the `content-key` has `content-type` 0x04, change the `content-type` to 0x02 and remove the `transaction-hash` field.
+  2.) Calculate `H(content-key)` where `H` denotes the SHA-256 hash function.
+
+Step 1 is done so that a `content-key` for a single transaction maps to the same `content-id` as the entire block body that contains the desired transaction. The individual transaction can then be retrieved and sent over the wire by itself. Otherwise the network would need to store each transaction twice: individually, as well as within the block body. This, along with the required inclusion proofs, would roughly triple the required network storage space.
```

Comment:
Just trying to make sure I understand this part: if we didn't remove the `transaction-hash` field then the content-key for a transaction and the block body containing the transaction would map to 2 different `content-id`s. If we are using the id as key into db, then this means we store the block body (which contains the tx) and the a sole tx object itself in the db, hence we store the transaction 2x, doubling storage space. Where are the "inclusion proofs" coming into play and adding extra storage space?

Reply:
The `Header.transactions_trie` represents a hexary patricia trie that contains all of the transactions in the block as values, with `rlp(transaction_index)` as the key into the trie.  There are two ways to prove that a transaction is part of a block.

1. Have the full ordered list of transactions for a block, construct this trie, and compare the state root.
2. Have a single transaction and a merkle proof against the trie, and validate the proof.

> Note that the trie in this case is not the ethereum state trie.  It's the same type of data structure, but it only contains the block transactions so it will be relatively small.

If we store individual transactions separate from the block body, then the DHT nodes storing the transactions will have to use option-2, also storing an inclusion proof, otherwise they will not be able to provide adequate proof to verify the returned data.  The storage overhead comes from having to both store the transaction twice (once in the block body, one in the stand-alone format) along with the inclusion proof.

For a sense of scale, the block bodies are 160GB and transactions represent the majority of this data.  My quick mental estimates suggest that we will incure 2x overhead meaning that we will need to store about 480GB of data if we store individual transactions.

Here is the breakdown of a geth node's storage

```
|    DATABASE     |      CATEGORY      |    SIZE    |   ITEMS   |
|-----------------|--------------------|------------|-----------|
| Key-Value store | Headers            | 50.19 MiB  |     90150 |
| Key-Value store | Bodies             | 6.66 GiB   |     90150 |
| Key-Value store | Receipt lists      | 3.93 GiB   |     90150 |
| Key-Value store | Difficulties       | 5.93 MiB   |    102805 |
| Key-Value store | Block number->hash | 4.96 MiB   |    102688 |
| Key-Value store | Block hash->number | 515.25 MiB |  13177531 |
| Key-Value store | Transaction index  | 14.89 GiB  | 444087861 |
| Key-Value store | Bloombit index     | 2.27 GiB   |   6591634 |
| Key-Value store | Contract codes     | 2.33 GiB   |    436305 |
| Key-Value store | Trie nodes         | 88.39 GiB  | 835099973 |
| Key-Value store | Trie preimages     | 547.13 KiB |      8893 |
| Key-Value store | Account snapshot   | 6.59 GiB   | 145109375 |
| Key-Value store | Storage snapshot   | 36.69 GiB  | 513200833 |
| Key-Value store | Clique snapshots   | 0.00 B     |         0 |
| Key-Value store | Singleton metadata | 819.00 B   |        12 |
| Ancient store   | Headers            | 5.73 GiB   |  13087382 |
| Ancient store   | Bodies             | 158.26 GiB |  13087382 |
| Ancient store   | Receipt lists      | 79.01 GiB  |  13087382 |
| Ancient store   | Difficulties       | 205.18 MiB |  13087382 |
| Ancient store   | Block number->hash | 474.28 MiB |  13087382 |
| Light client    | CHT trie nodes     | 0.00 B     |         0 |
| Light client    | Bloom trie nodes   | 0.00 B     |         0 |
|                         TOTAL        | 405.97 GIB |           |
```

Summing up our storage needs:

- trie nodes: 100GB
- trie leaf data: 50GB + unknown proof overhead (probably 2x upper bound)
- block bodies: 160GB
- singular transactions: 320GB (this is the additional overhead)
- receipts: 80GB
- TX Index: 50GB

Without singular transactions we need 440 GB of storage (ignoring replication).  This goes up to ~760GB which basically doubles our total storage needs.